### DISCUSSION TITLE
How to synchronize camera ?

### DISCUSSION THREAD

**Discussion by lchauvin** [Kitware/VolView#600] [Q&A]
Hello,

I would like to be able to synchronize cameras for all images (when I zoom on one image, and switch to another image, to get the same view).
I was able to do this with a previous version of VolView, but lots have changed since then.

I initially put a flag in CameraConfig to check if cameras should be sync or not, and then when a camera property was updated, I was checking the flag and updating all cameras with the same properties.

However, I don't think it works since I updated VolView, because it seems to sync CameraConfig with vtkCamera in usePersistentCameraConfig.ts, and vtkCamera doesn't have the flag I added, so it raised some error.

Could you suggest a way to do achieve this ?

Thank you.

**Comment by lchauvin** [comment]
I managed to synchronize camera properties, but when I change image, it resets the camera

**Comment by lchauvin** [comment]
Where would be the best place to put the logic for syncing cameras ? `usePersistCameraConfig.ts` ?

**Comment by lchauvin** [comment]
It seems from this code:

```ts
type KeyType = keyof CameraConfig;
  const keys: KeyType[] = [
    'position',
    'focalPoint',
    'viewUp',
    'parallelScale',
    'directionOfProjection',
  ];

  const cameraRefs = keys.reduce(
    (refs, key) => ({
      ...refs,
      [key]: guardedWritableRef(
        vtkFieldRef(camera, key),
        (incoming) => !!incoming
      ),
    }),
    {} as Record<KeyType, Ref<any>>
  );
```

that `CameraConfig` can only have properties that are in `vtkCamera`. What if I want to add a config parameter that is not in `vtkCamera` ?

How would I do that ?

**Reply by PaulHax** [reply]
I think if you change `keys` to 
```
const keys = [
    'position',
    'focalPoint',
    'viewUp',
    'parallelScale',
    'directionOfProjection',
  ] as const;
```
You should be back in biz.

Also, not 100% sure but I think the camera sync across view logic goes in a Pinia store like `view-configs/camera.ts`  Can leave `usePersistCameraConfig.ts` alone and let it focus on syncing Pinia store and VTK.js.

**Reply by lchauvin** [reply]
Won't this be an issue with that:

```
[key]: guardedWritableRef(
        vtkFieldRef(camera, key),
```

**Reply by PaulHax** [reply]
Yes.  I would `  type KeyType = (typeof keys)[number];`

**Reply by lchauvin** [reply]
When adding another property in `keys`, I'm getting this error at the `vtkFieldRef(camera, keys)`:

```
No overload matches this call.
  Overload 1 of 2, '(obj: MaybeRef<vtkCamera>, fieldName: "state" | "position" | "parallelScale" | "focalPoint" | "directionOfProjection" | "viewUp" | "roll" | "cameraLightTransformMatrix" | ... 40 more ... | "referenceByName"): Ref<...>', gave the following error.
    Argument of type '"position" | "parallelScale" | "focalPoint" | "directionOfProjection" | "viewUp" | "sync"' is not assignable to parameter of type '"state" | "position" | "parallelScale" | "focalPoint" | "directionOfProjection" | "viewUp" | "roll" | "cameraLightTransformMatrix" | "clippingRange" | "clippingRangeByReference" | ... 38 more ... | "referenceByName"'.
      Type '"sync"' is not assignable to type '"state" | "position" | "parallelScale" | "focalPoint" | "directionOfProjection" | "viewUp" | "roll" | "cameraLightTransformMatrix" | "clippingRange" | "clippingRangeByReference" | ... 38 more ... | "referenceByName"'.
  Overload 2 of 2, '(obj: MaybeRef<vtkCamera>, factory: GetterSetterFactory<unknown>): 
```

**Reply by lchauvin** [reply]
It seems to build though, so I don't know if it's something I could ignore.

**Reply by PaulHax** [reply]
Only need to add something to keys if you want it set on the VTK.js camera from the Pina store

**Reply by lchauvin** [reply]
What if I want to add something on the Pinia store that will not be set in VTK.js camera ? Because it seems both are using `CameraConfig`

**Reply by PaulHax** [reply]
Should be safe to remove all explcit reference to CameraConfig in usePersistCameraConfig.ts

**Reply by lchauvin** [reply]
I got this to work.

However, the issue I have now, is that, when I change image, it resets the camera.

I tried to comment out the `resetCamera` function in `SliceViewer` but it doesn't work.

I think it might have something to do with `disableAutoResetCamera` but I'm not sure. I tried to turn it off, but then the image is all black (or not in the FOV, not sure).

**Reply by PaulHax** [reply]
progress! ⭐️  

Not sure.  Are we re-creating the VTK.js scene/camera on switching images?  Are you saving the camera pose then applying after loading the matching image?

Perhaps time to open a PR and disscuss with code.

**Reply by lchauvin** [reply]
I created a PR here: https://github.com/Kitware/VolView/pull/615
